pkg/stash/with_gcs.go

```diff
@@ -0,0 +1,33 @@
+package stash
+
+import (
+	"context"
+
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/observ"
+)
+
+// WithGCSLock returns a distributed singleflight
+// using a GCS backend. See the config.toml documentation for details.
+func WithGCSLock(s Stasher) Stasher {
+	return &gcsLock{s}
+}
+
+type gcsLock struct {
+	stasher Stasher
+}
+
+func (s *gcsLock) Stash(ctx context.Context, mod, ver string) (newVer string, err error) {
+	const op errors.Op = "gcslock.Stash"
+	ctx, span := observ.StartSpan(ctx, op.String())
+	defer span.End()
+	newVer, err = s.stasher.Stash(ctx, mod, ver)
+	if err != nil {
+		// already been saved before, move on.
+		if errors.Is(err, errors.KindAlreadyExists) {
```

Comment:
@marwan-at-work
Can you walk me through a following scenario?
Let's say we have 2 Athens instances - a1 and a2.
a1 wants to stash mod@ver, it doesn't exist so it starts uploding it to the gcs. Now a2, 10ms later wants to stash mod@ver as well. When will a2 geht 412? Will it wait until a1 upload completed? Or will it return immediately?

Reply:
@marpio 
A2 will not wait for A1 to be finished. A2 will keep writing, and the moment you call `Close` it will check if there were previous writes. Keep in mind when two different clients write to the same Object path, they are not overriding each other, they both get a different "version" to write to. However, with the `ExactMatch` condition, the `Close` fails if there was a previous version already finished and closed. 

Therefore, if A2 finished before A1, then A2 is the *winner* and A1 will get a 412. Which is totally okay because from our perspective we don't care who finished first, we just care that a module is immutable and does not have "mixed" content from two different client.

It's basically a race to whoever finishes first. The `with_gcp_test.go` file ensures that because we launch 5 goroutines at the same tiem that all write to the same bucket with no guarantee to who finishes first. Therefore, the test check is that the content across all 3 files are consistent and that's what matters. 

I'm more than happy to listen to more scenarios or hear any concerns cuz this is def a very important piece of the Athens logic. Thanks!

Reply:
> It's basically a race to whoever finishes first.

Makes totaly sense now. Thanks!